Patch _fs_gs_helper when debugging windows remotely

AndersFelde opened the pull request:
I am using mingw64's gdbserver.exe to debug a windows PE remotely from Linux. Pwndbg will crash because the `fs_base` and `gs_base` registers does not exist when debugging in this way. By moving the `is_remote` check earlier (before it looks up the registers and crashes) pwndbg seems to work fine.

AndersFelde committed:
move remote check earlier

AndersFelde committed:
move remote check earlier

disconnect3d commented:
Hi, this is not a good solution: it will make this not work on linux gdbserver where it should work.

See https://github.com/pwndbg/pwndbg/issues/2467 for more info. What we should do is to check if fsbase/gsbase registers are there and fetch them only if they are there :).

disconnect3d commented:
@AndersFelde Can you fix this PR so we check the availability of the register?

AndersFelde committed:
add try except if register does not exist

AndersFelde commented:
@disconnect3d updated now. Is that a better approach, or do you want me to add try/except to the call inside _fs_gs_helper function?

disconnect3d commented:
@AndersFelde Instead of doing a try except, where we may miss one or another exception now (and miss bugs in the future) we should check if fsbase/gsbase registers are available before trying to fetch them.

AndersFelde committed:
check if register eists before trying to fetch

AndersFelde commented:
@disconnect3d what do you think now? I added the `regsiter_exists` function. It works when i test it on linux and windows

disconnect3d commented on pwndbg/gdblib/regs.py:
```diff
@@ -246,6 +255,9 @@ def _fs_gs_helper(self, regname: str, which: int) -> int:
         """Supports fetching based on segmented addressing, a la fs:[0x30].
         Requires ptrace'ing the child directory if i386."""
 
+        if not register_exists(regname):
```
Hmm I think this should be only executed for x86-64 case, so within the if that is in line 261.

disconnect3d replied:
This is probably why the tests on CI fail after this change 😅

disconnect3d reviewed (changes requested):
See comment - we need to move the check into the if. Also, pls fix lint issues (u cna run `./lint.sh -f` locally to autofix them)

AndersFelde committed:
moved register check to only x86-64

AndersFelde committed:
use .find() to determine wether register exists

AndersFelde commented:
@disconnect3d i've rewritten the check, and moved it inside the if-statement. When i run the tests locally with docker they all pass, but i see that some of them are not passing with github CI. I am also not able to see how the failing tests are related to my change at all?

disconnect3d commented:
@AndersFelde the test failures on CI are unrelated. They were uncovered by a recent change that made it so that XFAIL are accounted as test failures. I fixed like 6 of them but those 2 were left. I hope to find some time in upcomming days to fix them.. :)

patryk4815 commented:
@disconnect3d Please remember changing something in `gdblib` must be ported also to `aglib`
Right now we have differences, and this bug will reopen after removing gdblib xD
<img width="1130" alt="image" src="https://github.com/user-attachments/assets/ebe708ee-5c3e-49f1-87e3-e2128a2ae215">
